Actions/Custom Variables/Profiles overhaul

sebjulliand opened the pull request:
Fixes #2618
Fixes [#1451](https://github.com/codefori/vscode-ibmi/issues/1451)
## Changes
This PR introduce a new `Context` view as a replacement to the `Connection Profiles` view.
It is intended to enhance the user experience regarding the management of Actions, Custom Variables and Profiles.
<img width="464" height="153" alt="image" src="https://github.com/user-attachments/assets/1bb25b9c-6c2e-4f7d-90d5-1949c63225b2" />

It also introduces a new `CustomEditor` class that works like `CustomUI` but allows to open custom editors instead of custom panels.

### Actions
<img width="489" height="198" alt="image" src="https://github.com/user-attachments/assets/45d256cd-ae60-4b14-934a-da74fbbf634a" />

Actions are regrouped in three local categories, always displayed:
- Member actions
- Object actions
- Streamfile actions

Optionally, there will be additional categories for each workspace declaring local actions (in a local `.vscode/actions.json` file). These categories are named after the workspace they come from.

The actions available at the `Actions` node level are:
<img width="132" height="33" alt="image" src="https://github.com/user-attachments/assets/5d7f5f25-c8dd-4923-af40-1c94b26890da" />
- Create a new action
- Setup local actions
- Search actions
  - Enables the user to search for actions by their name and command. Each matching action is highlighted in yellow and its description shows `search match`.
  <img width="401" height="135" alt="image" src="https://github.com/user-attachments/assets/5a877895-b13b-413e-ac1e-c0c52f3232c8" />

- Go to next search result
  - Gives focus in the list on the next item that matched the search
- Clear search results

Creating a new action is also available on each `Action Type` node.
<img width="469" height="66" alt="image" src="https://github.com/user-attachments/assets/9d09169d-56b5-4907-872f-7c5cb14d08de" />

Clicking on an action opens the action editor. Making a change to the action will make the editor dirty and saving the action is done like in any editor, using `Save` (ctrl+s). The action panel was simply turned into an editor.
<img width="470" height="302" alt="image" src="https://github.com/user-attachments/assets/0a541d4a-6256-44df-801e-ed49ce14e078" />

Right-clicking on an action gives access to these options:
<img width="174" height="114" alt="image" src="https://github.com/user-attachments/assets/15038156-00c5-405c-b185-53c2ae4f413d" />

Standard actions and local actions are managed the same way in the UI; they are saved in the connection settings or in the local workspace file depending on their nature.

When an editor is open, an inline action is available on each action to run it on the active editor:
<img width="527" height="57" alt="image" src="https://github.com/user-attachments/assets/40382905-e82f-4618-82ca-9d1dab24a60e" />

Checking of the action can run on the active editor is done when the action is executed.


### Custom Variables
<img width="493" height="79" alt="image" src="https://github.com/user-attachments/assets/b2388f47-2c88-4726-b7d4-fc838b211695" />

Each variable declared in the current connection settings are listed here and can be created/changed/deleted from this view.
<img width="556" height="58" alt="image" src="https://github.com/user-attachments/assets/d4305fa3-31e6-4da3-b4eb-6b6a5ab53370" />
<img width="288" height="133" alt="image" src="https://github.com/user-attachments/assets/cedc1458-f47a-49ab-8148-5f97ddd43c23" />

Clicking on variable lets you change its value.

### Profiles
<img width="495" height="107" alt="image" src="https://github.com/user-attachments/assets/3a7c7f63-af6e-4186-99bb-b1e5d0c61d61" />

**Command profiles and Connection profiles have been merged - a process will run when the extension starts to create profiles out of existing Command profiles and show a dialog to warn the user the Command profiles have been migrated**.

Profiles can be created from the `Profiles` node using one of these inline actions:
<img width="369" height="32" alt="image" src="https://github.com/user-attachments/assets/71a01533-28f5-4811-a26b-5e832bafa6b2" />

- Create an empty profile
  - Asks for the name of the new profile then opens the profile editor.
- Create a new profile from the current settings
  - Asks for the name of the new profile then shows a notification to switch to it.
  <img width="460" height="91" alt="image" src="https://github.com/user-attachments/assets/93d7db07-0df1-419f-a02a-883958f914b2" />

Clicking on a profile opens the Profile editor that lets the user change some of the profile settings - object filters, IFS shortcuts and Custom variables are simply listed:
<img width="698" height="434" alt="image" src="https://github.com/user-attachments/assets/649b80af-1f36-4a46-bf47-06506fd87ba2" />
<img width="368" height="542" alt="image" src="https://github.com/user-attachments/assets/a487ad33-f23a-4d37-a0cf-a49cf539f7f1" />

Right-clicking on a profile gives access to these options:
<img width="177" height="125" alt="image" src="https://github.com/user-attachments/assets/934f0ea3-986c-4da8-9293-6f372876c39e" />

Activating a profile is done using the related inline action on the profile:
<img width="401" height="72" alt="image" src="https://github.com/user-attachments/assets/dd3a1022-f451-431f-b8bf-4181fdf2255f" />

Activating a profile will replace the current settings with the profile's. Before the settings are replaced, they are saved back into the profile they belong to, or into the default profile if no profile was active (default profile is a unique profile with no name). This allows to automatically save the profile state before it gets erased.

The active profile is shown in green in the list, and it can't be deleted:
<img width="328" height="110" alt="image" src="https://github.com/user-attachments/assets/56b7f1ae-2050-4017-b3ff-a98d5b0290c7" />

Its name is shown in the context description as well:
<img width="197" height="31" alt="image" src="https://github.com/user-attachments/assets/3045681d-6d8f-4428-872c-4493961c676f" />

Switching back to the default profile is done using the action that appears on the Profiles node inline actions:
<img width="420" height="58" alt="image" src="https://github.com/user-attachments/assets/a140c495-1d50-43d3-9ea5-ddc0409b0c5f" />

When a profile has a Library List Command (formerly a `Command Profile`), that command is ran and the library list is computed from it when the profile is activated. That operation can be re-run after the profile is made active by clicking the inline action on the profile item (the action only shows up if that profile has a library list command):
<img width="453" height="79" alt="image" src="https://github.com/user-attachments/assets/40bf150d-c307-476f-8f28-cc74afb8a21c" />


## How to test this PR

1. Create, edit, delete, rename, run actions
  a. Try generating local actions and make sure they are managed correctly in the local file
3. Create, edit, delete, rename variables
4. Create, edit, delete, rename, profiles
  a. Make sure custom variables are saved on the profile
  b. Make sure settings are correctly saved into the profile so they don't get lost when re-activating it
  c. Make sure the UI is consistent with the profile's content (views must be refreshed, etc)
5. Try actions described in the PR 😅
<!-- 
Example:
1. Run the test cases
6. Expand view A and right click on the node
7. Run 'Execute Thing' from the command palette
-->

## Checklist
<!-- Put an `x` in the relevant boxes -->
* [x] have tested my change
* [ ] updated relevant documentation

worksofliam commented on src/ui/diagnostics.ts:
```diff
@@ -78,7 +77,7 @@ export async function refreshDiagnosticsFromServer(instance: Instance, evfeventI
 export async function refreshDiagnosticsFromLocal(instance: Instance, evfeventInfo: EvfEventInfo) {
   if (evfeventInfo.workspace) {
-    const evfeventFiles = await getEvfeventFiles(evfeventInfo.workspace);
+    const evfeventFiles = await vscode.workspace.findFiles(new vscode.RelativePattern(evfeventInfo.workspace, `**/.evfevent/*`), null);
```
Why this change

chrjorgensen reviewed (commented):
@sebjulliand As you know, I'm currently traveling, so I haven't got the time for a thorough review but just played around a little.

Nice work, which will simplify the concept of profiles. 👍 

Two comments:
1. After creating a new profile and copying a command, the context view is not refreshed and I didn't see the new command until a manual refresh.
2. Could we have the active profile attached to the connection name in the status bar, like "System (profile)"?

SanjulaGanepola commented on package.json:
```diff
@@ -1708,6 +1614,160 @@
+				"icon": "$(search-stop)"
+			},
+			{
+				"command": "code-for-ibmi.context.action.create",
```
When using `Create action` on a workspace in the `Context` view, I get this error:
<img width="2927" height="1028" alt="image" src="https://github.com/user-attachments/assets/50d67e47-bad5-4505-8ec2-da97f6623001" />

sebjulliand replied:
And fixed 👌🏻

SanjulaGanepola commented:
Shall we also link these issues:
- https://github.com/codefori/vscode-ibmi/issues/1451
- https://github.com/codefori/vscode-ibmi/issues/2618

Looks like you have most of those requests covered here in this PR!

sebjulliand commented:
Back at you folks, I updated the icons and fixed the reported issues.
I also added additional checks so actions and profiles can't be renamed/deleted while being edited.

SanjulaGanepola commented:
Other ideas:
1. When we do a `Launch Workspace Actions Setup...`, can we auto refresh the view so it loads the new actions?
2. When we get the `Local actions are already defined for this workspace. Do you want to overwrite them?` message, can we get an option to `Append`?

SanjulaGanepola commented on src/ui/Tools.ts:
```diff
@@ -199,7 +199,23 @@ export namespace VscodeTools {
 
+  export function profileToToolTip(profile: ConnectionProfile) {
+    const tooltip = new MarkdownString(generateTooltipHtmlTable('', {
+      "Home directory": profile.homeDirectory,
```
```suggestion
      "Home Directory": profile.homeDirectory,
```

SanjulaGanepola reviewed (approved):
Great work!!
